### PR TITLE
add filter to cli command, resolve profile related to volume group and add Listwithfilter method to volume group.

### LINKED ISSUE #294: Add authentication and request filter in osdsctl and client package
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

**What happened**:
According to keystone design (#289) and API request filter support (#284), opensds api-server has supported authentication and request filter features, but as for client and cli tool, these features also need to be supported.

**What you expected to happen**:
Add keystone authentication and api request filter support in cli and client tool.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Hotpot(release/branch) version:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by lasshpanzer**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #294 

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by leonwanghui** [comment]
Because this patch will have impact on Client package, so we need to have a detailed discussion.

File: client/client.go [hunk 1]
```diff
@@ -117,3 +120,31 @@ func (c *Client) Reset() *Client {
 	c = &Client{}
 	return c
 }
+
+func processListPara(args []interface{}) (string, error) {
```
Comment (xxwjj) [inline client/client.go]: s/para/param

File: client/client.go [hunk 2]
```diff
@@ -117,3 +120,31 @@ func (c *Client) Reset() *Client {
...
+		}
+		filter = args[0].(map[string]string)
+	}
+
+	if filter != nil {
+		for k, v := range filter {
+			if v == "" {
+				continue
+			}
+			urlpara = append(urlpara, k+"="+v)
```
Comment (xxwjj) [inline client/client.go]: same

File: client/volume.go [hunk 3]
```diff
@@ -184,16 +192,23 @@ func (v *VolumeMgr) GetVolumeAttachment(atcID string) (*model.VolumeAttachmentSp
...
 // ListVolumeAttachments
-func (v *VolumeMgr) ListVolumeAttachments() ([]*model.VolumeAttachmentSpec, error) {
-	var res []*model.VolumeAttachmentSpec
+func (v *VolumeMgr) ListVolumeAttachments(args ...interface{}) ([]*model.VolumeAttachmentSpec, error) {
 	url := strings.Join([]string{
 		v.Endpoint,
 		urls.GenerateAttachmentURL(urls.Client, v.TenantId)}, "/")
 
-	if err := v.Recv(url, "GET", nil, &res); err != nil {
+	para, err := processListPara(args)
```
Comment (xxwjj) [inline client/volume.go]: s/para/param

File: osdsctl/cli/volume.go [hunk 4]
```diff
@@ -76,7 +76,41 @@ var (
...
+	volLimit     string
+	volOffset    string
+	volSortDir   string
+	volSortKey   string
+	volId        string
+	volCreatedAt string
+	volUpdatedAt string
+	volTenantId  string
+	volUserId    string
+	volSize      string
```
Comment (xxwjj) [inline osdsctl/cli/volume.go]: why not the type of volSize is string rather than int64.

File: openapi-spec/swagger.yaml [hunk 5]
```diff
@@ -748,6 +746,17 @@ paths:
             type: array
             items:
               $ref: '#/definitions/VolumeGroup'
+          examples:
+            application/json:
+              - id: 015184f3-8e73-47fd-8f57-26ea912e2a6b
+                name: volumeGroup-demo
+                stauts: available
```
Comment (xing-yang) [inline openapi-spec/swagger.yaml]: typo: status

File: osdsctl/cli/dock.go [hunk 6]
```diff
@@ -43,9 +43,40 @@ var dockListCommand = &cobra.Command{
...
+	dockDriverName  string
+)
+
 func init() {
+	dockListCommand.Flags().StringVarP(&dockLimit, "limit", "", "50", "the number of ertries displayed per page")
+	dockListCommand.Flags().StringVarP(&dockOffset, "offset", "", "0", "all requested data offsets")
+	dockListCommand.Flags().StringVarP(&dockSortDir, "sortDir", "", "desc", "the sort direction of all requested data. supports asc or desc(default)")
+	dockListCommand.Flags().StringVarP(&dockSortKey, "sortKey", "", "id", "the sort key of all requested data. supports id(default), name, status, endpoint, drivername, description")
+	dockListCommand.Flags().StringVarP(&dockId, "id", "", "", "list docks by id")
+	dockListCommand.Flags().StringVarP(&dockCreatedAt, "createdAt", "", "", "list docks by created time")
```
Comment (xing-yang) [inline osdsctl/cli/dock.go]: Does this list docks by the exact created time?  It is more useful to list docks created before or after a timestamp.

Same comments applied to other timestamp filters.

File: osdsctl/cli/pool.go [hunk 7]
```diff
@@ -42,7 +42,41 @@ var poolListCommand = &cobra.Command{
...
+	poolListCommand.Flags().StringVarP(&poolCreatedAt, "createdAt", "", "", "list pools by created time")
+	poolListCommand.Flags().StringVarP(&poolName, "name", "", "", "list pools by name")
+	poolListCommand.Flags().StringVarP(&poolUpdatedAt, "updatedAt", "", "", "list pools by updated time")
+	poolListCommand.Flags().StringVarP(&poolDescription, "description", "", "", "list pools by description")
+	poolListCommand.Flags().StringVarP(&poolStatus, "status", "", "", "list pools by status")
+	poolListCommand.Flags().StringVarP(&poolStorageType, "storageType", "", "", "list pools by storage type")
+	poolListCommand.Flags().StringVarP(&poolDockId, "dockId", "", "", "list pools by dock id")
+	poolListCommand.Flags().StringVarP(&poolAvailabilityZone, "availabilityZone", "", "", "list pools by availability zone")
+	poolListCommand.Flags().StringVarP(&poolTotalCapacity, "totalCapacity", "", "", "list pools by totalCapacity")
+	poolListCommand.Flags().StringVarP(&poolFreeCapacity, "freeCapacity", "", "", "list pools by freeCapacity")
```
Comment (xing-yang) [inline osdsctl/cli/pool.go]: Similarly for capacity, it is fine to list pools == some capacity, but it is more useful to list pools >= or/and <= some capacity.

File: pkg/api/db.go [hunk 8]
```diff
@@ -187,7 +187,9 @@ func CreateVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) (
 }
 
 func DeleteVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) error {
-	if in.Status != model.VolumeSnapAvailable {
+	invalidStatus := []string{model.VolumeSnapAvailable, model.VolumeSnapError,
```
Comment (xxwjj) [inline pkg/api/db.go]: If one snapshot is in model.VolumeSnapAvailable and model.VolumeSnapError, model.VolumeSnapErrorDeleting, we could delete it.

Comment (xxwjj) [reply pkg/api/db.go]: And when one snapshot in creatint status, we should not delete it.

File: pkg/api/db.go [hunk 9]
```diff
@@ -203,7 +205,7 @@ func DeleteVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) e
 //Just modify the state of the volume to be deleted in the DB, the real deletion in another thread
 func DeleteVolumeDBEntry(ctx *c.Context, in *model.VolumeSpec) error {
 	invalidStatus := []string{model.VolumeAvailable, model.VolumeError,
-		model.VolumeErrorDeleting, model.VolumeErrorExtending}
+		model.VolumeErrorDeleting, model.VolumeErrorExtending, model.VolumeDeleted}
```
Comment (xxwjj) [inline pkg/api/db.go]: Add volumeCreating and volumeDeleting statuses and remove volumeErrorDeleting status.

File: client/client.go (hunk 2 above)
Comment (leonwanghui) [reply client/client.go]: s/urlpara/urlParam

File: osdsctl/cli/dock.go [hunk 10]
```diff
@@ -43,9 +44,40 @@ var dockListCommand = &cobra.Command{
...
+	dockUpdatedAt   string
+	dockDescription string
+	dockStatus      string
+	dockStorageType string
+	dockEndpoint    string
+	dockDriverName  string
+)
+
 func init() {
+	dockListCommand.Flags().StringVarP(&dockLimit, "limit", "", "50", "the number of ertries displayed per page")
```
[the remark this answers is not in the document]
Comment (lasshpanzer) [reply osdsctl/cli/dock.go]: For example?

[the remark this answers is not in the document]
Comment (lasshpanzer) [reply osdsctl/cli/dock.go]: Show me which parameter is no need to add.

File: osdsctl/cli/volume.go (hunk 4 above)
Comment (lasshpanzer) [reply osdsctl/cli/volume.go]: String in cobra is more better handling.

File: pkg/api/db.go (hunk 8 above)
Comment (lasshpanzer) [reply pkg/api/db.go]: What do you want to say?

File: osdsctl/cli/pool.go (hunk 7 above)
Comment (lasshpanzer) [reply osdsctl/cli/pool.go]: Only supports accurate query or not support this parameter. Unlike SQL statements, if you want to achieve this function, you need to achieve a very very complex implementation

File: osdsctl/cli/dock.go (hunk 6 above)
Comment (lasshpanzer) [reply osdsctl/cli/dock.go]: Only supports accurate query or not support this parameter. Unlike SQL statements, if you want to achieve this function, you need to achieve a very very complex implementation

File: client/client.go (hunk 2 above)
Comment (lasshpanzer) [reply client/client.go]: ok

File: client/client.go (hunk 1 above)
Comment (lasshpanzer) [reply client/client.go]: ok

File: openapi-spec/swagger.yaml (hunk 5 above)
Comment (lasshpanzer) [reply openapi-spec/swagger.yaml]: ok

File: osdsctl/cli/dock.go (hunk 6 above)
Comment (xing-yang) [reply osdsctl/cli/dock.go]: Please remove the filtering using timestamps such as createdAt and updatedAt.  We can look at filtering by a range later.

Comment (lasshpanzer) [reply osdsctl/cli/dock.go]: Got that

**Pull request closed without merging**